lib/Dialect/Pipeline/Transforms/ScheduleLinearPipeline.cpp

```diff
@@ -0,0 +1,140 @@
+//===- ScheduleLinearPipeline.cpp - Linear pipeline scheduling pass -----*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions linear pipeline scheduling pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Analysis/DependenceAnalysis.h"
+#include "circt/Analysis/SchedulingAnalysis.h"
+#include "circt/Scheduling/Algorithms.h"
+#include "circt/Scheduling/Utilities.h"
+#include "circt/Support/BackedgeBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "pipeline-schedule-linear"
+
+using namespace mlir;
+using namespace circt;
+using namespace pipeline;
+
+namespace {
+
+class ScheduleLinearPipelinePass
+    : public ScheduleLinearPipelineBase<ScheduleLinearPipelinePass> {
+public:
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+// Returns true if 'op' should be ignored in the scheduling problem.
+static bool ignoreOp(Operation *op) {
+  return isa<hw::ConstantOp, pipeline::ReturnOp>(op);
+}
+
+void ScheduleLinearPipelinePass::runOnOperation() {
+  auto pipeline = getOperation();
+
+  auto stageOpIt = pipeline.getOps<PipelineStageOp>();
+  auto stageRegOpIt = pipeline.getOps<PipelineStageRegisterOp>();
+
+  if ((std::distance(stageOpIt.begin(), stageOpIt.end()) != 0) ||
+      (std::distance(stageRegOpIt.begin(), stageRegOpIt.end()) != 0)) {
+    pipeline.emitError("Pipeline cannot have any stages or stage registers.");
+    return signalPassFailure();
+  }
+
+  // Load operator info from attribute.
+  auto problem = SharedOperatorsProblem::get(pipeline);
```

Comment:
For this use-case, couldn't this be just a Problem? I don't see it using any features of a shared operators problem.

Reply:
As mentioned in the PR message, i didn't add the shared operator "support" yet. To me, that would be a valuable addition to the design space that this scheduling pass can explore. But in reality, we most likely do not need it for our initial work in this space - @teqdruid thoughts?

Reply:
I see, I guess I'd recommend starting with the simplest problem type for now, since it shouldn't be hard to update once you are ready to fill in the shared operator properties, or whatever else.

Reply:
@mortbopet yes, that is correct -- we won't need it initially.